File: src/market-structure.js
import {
  DEFAULT_PIVOT_STRENGTH,
  DEFAULT_STRUCTURE_LOOKBACK,
  DEFAULT_TRENDLINE_PIVOTS,
} from "./constants.js";

function analyzeMarketStructure(candles, settings = {}) {
  if (candles.length === 0) {
    return emptyMarketStructure();
  }

  const latestIndex = candles.length - 1;
  const latestClose = candles[latestIndex].close;
  const lookback = Math.max(5, settings.structureLookback ?? DEFAULT_STRUCTURE_LOOKBACK);
  const pivotStrength = Math.max(1, settings.pivotStrength ?? DEFAULT_PIVOT_STRENGTH);
  const trendlinePivotCount = Math.max(
    2,
    settings.trendlinePivotCount ?? DEFAULT_TRENDLINE_PIVOTS,
  );
  const startIndex = Math.max(0, candles.length - lookback);
  const recentCandles = candles.slice(startIndex);
  const pivots = findSwingPivots(candles, pivotStrength).filter(
    (pivot) => pivot.index >= startIndex,
  );

  const swingHighs = pivots.filter((pivot) => pivot.type === "high");
  const swingLows = pivots.filter((pivot) => pivot.type === "low");

  return {
    support: nearestLevelBelow(
      latestClose,
      swingLows.map((pivot) => pivot.price),
      recentCandles.map((candle) => candle.low),
    ),
    resistance: nearestLevelAbove(
      latestClose,
      swingHighs.map((pivot) => pivot.price),
      recentCandles.map((candle) => candle.high),
    ),
    supportTrendline: projectedTrendlineValue(swingLows, latestIndex, trendlinePivotCount),
    resistanceTrendline: projectedTrendlineValue(swingHighs, latestIndex, trendlinePivotCount),
  };
}

function emptyMarketStructure() {
  return {
    support: null,
    resistance: null,
    supportTrendline: null,
    resistanceTrendline: null,
  };
}

function findSwingPivots(candles, strength) {
  const pivots = [];

  for (let index = strength; index < candles.length - strength; index += 1) {
    const candle = candles[index];
    let isSwingHigh = true;
    let isSwingLow = true;

    for (let offset = 1; offset <= strength; offset += 1) {
      if (
        candle.high <= candles[index - offset].high ||
        candle.high <= candles[index + offset].high
      ) {
        isSwingHigh = false;
      }
      if (
        candle.low >= candles[index - offset].low ||
        candle.low >= candles[index + offset].low
      ) {
        isSwingLow = false;
      }
    }

    if (isSwingHigh) {
      pivots.push({ type: "high", index, price: candle.high });
    }
    if (isSwingLow) {
      pivots.push({ type: "low", index, price: candle.low });
    }
  }

  return pivots;
}

function nearestLevelBelow(price, preferredLevels, fallbackLevels) {
  const below = preferredLevels.filter((level) => level <= price);
  if (below.length > 0) {
    return Math.max(...below);
  }
  if (fallbackLevels.length === 0) {
    return null;
  }
  return Math.min(...fallbackLevels);
}

function nearestLevelAbove(price, preferredLevels, fallbackLevels) {
  const above = preferredLevels.filter((level) => level >= price);
  if (above.length > 0) {
    return Math.min(...above);
  }
  if (fallbackLevels.length === 0) {
    return null;
  }
  return Math.max(...fallbackLevels);
}

function projectedTrendlineValue(pivots, targetIndex, pivotCount) {
  const selected = pivots.slice(-pivotCount);
  if (selected.length < 2) {
    return null;
  }

  const averageIndex = selected.reduce((total, pivot) => total + pivot.index, 0) / selected.length;
  const averagePrice = selected.reduce((total, pivot) => total + pivot.price, 0) / selected.length;
  const variance = selected.reduce(
    (total, pivot) => total + (pivot.index - averageIndex) ** 2,
    0,
  );
  if (variance === 0) {
    return null;
  }

  const covariance = selected.reduce(
    (total, pivot) => total + (pivot.index - averageIndex) * (pivot.price - averagePrice),
    0,
  );
  const slope = covariance / variance;
  const intercept = averagePrice - slope * averageIndex;
  return slope * targetIndex + intercept;
}

export { analyzeMarketStructure };

File: src/aurex/market_structure.py
from aurex.constants import (
    DEFAULT_PIVOT_STRENGTH,
    DEFAULT_STRUCTURE_LOOKBACK,
    DEFAULT_TRENDLINE_PIVOTS,
)


def analyze_market_structure(candles, settings=None):
    if len(candles) == 0:
        return empty_market_structure()

    settings = settings or {}
    latest_index = len(candles) - 1
    latest_close = candles[latest_index]["close"]
    lookback = max(5, _setting(settings, "structure_lookback", DEFAULT_STRUCTURE_LOOKBACK))
    pivot_strength = max(1, _setting(settings, "pivot_strength", DEFAULT_PIVOT_STRENGTH))
    trendline_pivot_count = max(
        2,
        _setting(settings, "trendline_pivot_count", DEFAULT_TRENDLINE_PIVOTS),
    )
    start_index = max(0, len(candles) - lookback)
    recent_candles = candles[start_index:]
    pivots = [
        pivot
        for pivot in find_swing_pivots(candles, pivot_strength)
        if pivot["index"] >= start_index
    ]

    swing_highs = [pivot for pivot in pivots if pivot["type"] == "high"]
    swing_lows = [pivot for pivot in pivots if pivot["type"] == "low"]

    return {
        "support": nearest_level_below(
            latest_close,
            [pivot["price"] for pivot in swing_lows],
            [candle["low"] for candle in recent_candles],
        ),
        "resistance": nearest_level_above(
            latest_close,
            [pivot["price"] for pivot in swing_highs],
            [candle["high"] for candle in recent_candles],
        ),
        "supportTrendline": projected_trendline_value(
            swing_lows,
            latest_index,
            trendline_pivot_count,
        ),
        "resistanceTrendline": projected_trendline_value(
            swing_highs,
            latest_index,
            trendline_pivot_count,
        ),
    }


def empty_market_structure():
    return {
        "support": None,
        "resistance": None,
        "supportTrendline": None,
        "resistanceTrendline": None,
    }


def find_swing_pivots(candles, strength):
    pivots = []

    for index in range(strength, len(candles) - strength):
        candle = candles[index]
        is_swing_high = True
        is_swing_low = True

        for offset in range(1, strength + 1):
            previous_high = candles[index - offset]["high"]
            next_high = candles[index + offset]["high"]
            previous_low = candles[index - offset]["low"]
            next_low = candles[index + offset]["low"]

            if candle["high"] <= previous_high or candle["high"] <= next_high:
                is_swing_high = False
            if candle["low"] >= previous_low or candle["low"] >= next_low:
                is_swing_low = False

        if is_swing_high:
            pivots.append({"type": "high", "index": index, "price": candle["high"]})
        if is_swing_low:
            pivots.append({"type": "low", "index": index, "price": candle["low"]})

    return pivots


def nearest_level_below(price, preferred_levels, fallback_levels):
    below = [level for level in preferred_levels if level <= price]
    if len(below) > 0:
        return max(below)
    if len(fallback_levels) == 0:
        return None
    return min(fallback_levels)


def nearest_level_above(price, preferred_levels, fallback_levels):
    above = [level for level in preferred_levels if level >= price]
    if len(above) > 0:
        return min(above)
    if len(fallback_levels) == 0:
        return None
    return max(fallback_levels)


def projected_trendline_value(pivots, target_index, pivot_count):
    selected = pivots[-pivot_count:]
    if len(selected) < 2:
        return None

    average_index = sum(pivot["index"] for pivot in selected) / len(selected)
    average_price = sum(pivot["price"] for pivot in selected) / len(selected)
    variance = sum((pivot["index"] - average_index) ** 2 for pivot in selected)
    if variance == 0:
        return None

    covariance = sum(
        (pivot["index"] - average_index) * (pivot["price"] - average_price)
        for pivot in selected
    )
    slope = covariance / variance
    intercept = average_price - slope * average_index
    return slope * target_index + intercept


def _setting(settings, field_name, fallback):
    if isinstance(settings, dict):
        return settings.get(field_name, fallback)
    return getattr(settings, field_name, fallback)
